File: backend/src/main/kotlin/com/alpha/service/RegionService.kt
package com.alpha.service

import com.alpha.domain.entity.RegionEntity
import com.alpha.domain.repository.RegionRepository
import com.alpha.service.dto.*
import com.alpha.service.exception.ConflictException
import com.alpha.service.exception.NotFoundException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class RegionService(
    private val regionRepository: RegionRepository
) {
    
    fun getAllRegions(): List<RegionResponse> {
        val regions = regionRepository.findByIsActiveTrue()
        return regions.map { mapToRegionResponse(it) }
    }
    
    fun getRegion(regionId: UUID): RegionResponse {
        val region = regionRepository.findById(regionId)
            .orElseThrow { NotFoundException("Region not found") }
        
        return mapToRegionResponse(region)
    }
    
    fun getRegionByCode(code: String): RegionResponse {
        val region = regionRepository.findByCode(code)
            ?: throw NotFoundException("Region not found")
        
        return mapToRegionResponse(region)
    }
    
    fun createRegion(request: CreateRegionRequest): RegionResponse {
        // Check if code already exists
        if (regionRepository.existsByCode(request.code)) {
            throw ConflictException("Region with this code already exists")
        }
        
        val region = RegionEntity().apply {
            name = request.name
            code = request.code
            country = request.country
            timezone = request.timezone
            isActive = request.isActive
        }
        
        val savedRegion = regionRepository.save(region)
        return mapToRegionResponse(savedRegion)
    }
    
    fun updateRegion(regionId: UUID, request: UpdateRegionRequest): RegionResponse {
        val region = regionRepository.findById(regionId)
            .orElseThrow { NotFoundException("Region not found") }
        
        request.name?.let { region.name = it }
        request.code?.let { 
            // Check if new code already exists (if different from current)
            if (it != region.code && regionRepository.existsByCode(it)) {
                throw ConflictException("Region with this code already exists")
            }
            region.code = it
        }
        request.country?.let { region.country = it }
        request.timezone?.let { region.timezone = it }
        request.isActive?.let { region.isActive = it }
        
        val updatedRegion = regionRepository.save(region)
        return mapToRegionResponse(updatedRegion)
    }
    
    fun deleteRegion(regionId: UUID) {
        val region = regionRepository.findById(regionId)
            .orElseThrow { NotFoundException("Region not found") }
        
        // Soft delete by marking as inactive
        region.isActive = false
        regionRepository.save(region)
    }
    
    fun searchRegions(search: String): List<RegionResponse> {
        val regions = regionRepository.search(search)
        return regions.map { mapToRegionResponse(it) }
    }
    
    private fun mapToRegionResponse(region: RegionEntity): RegionResponse {
        return RegionResponse(
            id = region.requiredId,
            name = region.name,
            code = region.code,
            country = region.country,
            timezone = region.timezone,
            isActive = region.isActive,
            createdAt = region.createdAt!!
        )
    }
}

File: backend-scala-zio/src/main/scala/com/alpha/service/RegionService.scala
package com.alpha.service

import zio.*
import com.alpha.repository.*
import com.alpha.domain.model.*
import com.alpha.dto.*
import com.alpha.provider.*
import java.util.UUID

trait RegionService:
  def getAllRegions: Task[List[Region]]
  def getRegionById(id: UUID): Task[Option[Region]]
  def getRegionByCode(code: String): Task[Option[Region]]
  def searchRegions(query: String): Task[List[Region]]
  def createRegion(req: CreateRegionRequest): Task[Region]
  def updateRegion(id: UUID, req: UpdateRegionRequest): Task[Region]
  def deleteRegion(id: UUID): Task[Unit]

object RegionService:
  val layer: ZLayer[RegionRepository & TimeProvider & UUIDProvider, Nothing, RegionService] =
    ZLayer.fromFunction(new RegionServiceImpl(_, _, _))

class RegionServiceImpl(regionRepo: RegionRepository, timeProvider: TimeProvider, uuidProvider: UUIDProvider)
  extends RegionService:

  override def getAllRegions: Task[List[Region]] =
    regionRepo.findAll

  override def getRegionById(id: UUID): Task[Option[Region]] =
    regionRepo.findById(id)

  override def getRegionByCode(code: String): Task[Option[Region]] =
    regionRepo.findByCode(code)

  override def searchRegions(query: String): Task[List[Region]] =
    regionRepo.search(query)

  override def createRegion(req: CreateRegionRequest): Task[Region] =
    for
      exists <- regionRepo.existsByCode(req.code)
      _      <- ZIO.fail(new Exception("Region with this code already exists")).when(exists)
      region  = Region(
                  id = uuidProvider.randomUUID(),
                  name = req.name,
                  code = req.code,
                  country = req.country,
                  timezone = req.timezone,
                  isActive = true,
                  createdAt = timeProvider.now(),
                  updatedAt = None
                )
      id     <- regionRepo.create(region)
    yield region.copy(id = id)

  override def updateRegion(id: UUID, req: UpdateRegionRequest): Task[Region] =
    for
      regionOpt <- regionRepo.findById(id)
      region    <- ZIO.fromOption(regionOpt).orElseFail(new Exception("Region not found"))
      updated    = region.copy(
                     name = req.name.getOrElse(region.name),
                     code = req.code.getOrElse(region.code),
                     country = req.country.getOrElse(region.country),
                     timezone = req.timezone.getOrElse(region.timezone),
                     isActive = req.isActive.getOrElse(region.isActive),
                     updatedAt = Some(timeProvider.now())
                   )
      _         <- regionRepo.update(updated)
    yield updated

  override def deleteRegion(id: UUID): Task[Unit] =
    regionRepo.delete(id).unit
